### PR TITLE
NPCs: don't take e-cash when trading

### LINKED ISSUE #27449: [RFC] Post Cataclysm trade and quick and dirty faction cash
### Ecash considered harmful
NPCs accept pre-Cataclysm electronic cash.  This makes a huge number of features and missions related to the Free Merchants, Evac Center, and Refugee Ranch/Tacoma Commune fairly pointless because no sane survivor is going to trade meat jerky at $3/piece, or send NPC allies out to work at $8/hour, when they can go kill some random zombie and pick up a cash card worth $300.

In order to make post-Cataclysm trade sensible, NPCs need to heavily discount eCash and require PCs to engage in barter.

### Faction cash
Actual barter is a hassle, and requiring NPCs to have trade-goods for everything the PCs wants to buy or sell is going to quickly run into limitations with the NPC code.  Instead, we should abstract some kind of faction cash, and let NPCs buy and sell stuff using faction cash.  Different factions would discount cash from other factions at different amounts: full value or nearly so for friendly factions (the Old Guard and the Free Merchants), much less for hostile factions (Free Merchants are going to value Hells Raider cash at 10% or less).  All factions discount eCash at 99.9%, so $1000 in eCash buys $1 of goods from the Free Merchants.

Implementation wise, the current `int character::cash` gets replaced with `std::map<faction_id, int> character::cash`, and factions get a similar map containing their discount rates (probably a fixed point value with 6 sigfigs, so 100,000,000 represents 100% and 100,000 represents 0.1%).

When the player attempts to trade with an NPC, his current cash map gets discounted and summed into a single cash value in the npc trade window.  Purchases are made in descending order from the least discounted faction cash to the worst (or with the value of a bartered good).

### Ramifications
Some costs at the Free Merchants would need to be adjusted in lines with the new costs.  The Refugee Ranch doctor charging $5000 to install a CBM is reasonable when that represents 30-40 zombies killed and looted, but it may be too much when it represents 1800 pieces of meat jerky.

See also  #27447. and #12539

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Features "NPCs: don't take e-cash when trading"```
#### Purpose of change
Fixes #27449
NPCs will no longer sell you valuable goods in exchange for electronic cash you skimmed off the cash cards you got by killing a bunch of zombies.  NPCs will trade you goods for goods.  NPCs in a faction with a currency will treat that currency as a medium of exchange, always trading for 100% of value regardless of need, supply, or relative trader skill.

#### Describe the solution
Add a debug menu option to alter an NPC's op_of_u, because it was quick to write and I needed to edit NPC opinions.

The avatar's cash is no longer counted when trading with an NPC.  All trades are barter only.  NPCs in a faction with a currency will give full value when trading that currency, regardless of the NPC's needs or supplies or the relative skill (this may not always been in the avatar's favor, and I think there's a possibility for arbitrage but I'm not going to stop that from happening).  NPCs will allow an avatar to owe them a small amount and still complete the trade, depending on how much the NPC likes the avatar and the NPC's altruism.

Add a namespace for NPC trading and move some random functions into it.

Revise a couple of dialogue conditionals and effects to handle the new system.  For instance, bulk trades get paid in faction currency, and other NPC services require that either the NPC owe the player the cost of service OR that the avatar trade away enough stuff to complete the trade.  The scavenger merc in the refugee center now requires 2500 Merch instead of $8000 and explains (somewhat incorrectly) what a Merch is.

#### Describe alternatives you've considered
Probably could be refactored again in light of #31289 but it stands fine on its own.

#### Additional context
no_ecash_bulk_trade
Selling meat jerky for Merch

e_cash_trading_with_debt
The NPC likes me, and let me owe her for the scissors, so I'm ahead at the start of this trade.

e_cash_trading_need_more_money
but now I need to satisfy the old debt and the mop price because I'm past my credit limit

e_cash_trading_debt_satisfied
my switchblade will do the job; the NPC will owe me a little.
note in all 3 instances the Merch are worth $3.20, regardless of relative skill.

### CODE REVIEW

**Comment by I-am-Erk** [comment]
We're still faced with the annoying issue that you can only barter a single merch per transaction, unfortunately? Otherwise this is excellent and I'm excited to see it happening

**Comment by mlangsdorf** [comment]
Yeah, I'm working on fixing that and will append the commit to this series.

**Comment by I-am-Erk** [comment]
Oh awesome, you're a gentleman and a scholar.

File: src/npctrade.h [hunk 1]
```diff
@@ -30,12 +30,18 @@ struct item_pricing {
...
+bool pay_npc( npc &np, int cost );
+
 int cash_to_favor( const npc &, int cash );
 
 inventory inventory_exchange( inventory &inv,
                               const std::set<item *> &without, const std::vector<item *> &added );
 std::vector<item_pricing> init_selling( npc &p );
 std::vector<item_pricing> init_buying( npc &p, player &u );
 bool trade( npc &p, int cost, const std::string &deal );
+};
```
Comment (kevingranade) [inline src/npctrade.h]: Pedantic warning about unnecessary ;
```suggestion
}
```

File: src/npctalk.cpp [hunk 2]
```diff
@@ -1757,6 +1757,10 @@ void talk_effect_fun_t::set_u_buy_item( const std::string &item_name, int cost,
     function = [item_name, cost, count, container_name]( const dialogue & d ) {
         npc &p = *d.beta;
         player &u = *d.alpha;
+        if( !npc_trading::pay_npc( p, cost ) ) {
+            popup( "You can't afford it!" );
```
Comment (Night-Pryanik) [inline src/npctalk.cpp]:
```suggestion
            popup( _( "You can't afford it!" ) );
```

File: src/npctalk.cpp [hunk 3]
```diff
@@ -2008,8 +2011,34 @@ void talk_effect_fun_t::set_bulk_trade_accept( bool is_trade, bool is_npc )
...
+                if( d.beta->value( pay ) > 0 ) {
+                    int required = price / d.beta->value( pay );
+                    int buyer_has = required;
+                    if( is_npc ) {
+                        buyer_has = std::min( buyer_has, buyer->charges_of( pay_in ) );
+                        buyer->use_charges( pay_in, buyer_has );
+                    } else {
+                        if( buyer_has == 1 ) {
+                            //~ %1%s is the NPC name, %2$s is an item
+                            popup( _( "%1$s gives you a %2$s" ), d.beta->disp_name(),
```
Comment (Night-Pryanik) [inline src/npctalk.cpp]:
```suggestion
                            popup( _( "%1$s gives you a %2$s." ), d.beta->disp_name(),
```

File: src/npctalk.cpp [hunk 4]
```diff
@@ -2008,8 +2011,34 @@ void talk_effect_fun_t::set_bulk_trade_accept( bool is_trade, bool is_npc )
...
+                                   pay.tname() );
+                        } else if( buyer_has > 1 ) {
+                            //~ %1%s is the NPC name, %2$d is a number of items, %3$s are items
+                            popup( _( "%1$s gives you %2$d %3$s" ), d.beta->disp_name(), buyer_has,
```
Comment (Night-Pryanik) [inline src/npctalk.cpp]:
```suggestion
                            popup( _( "%1$s gives you %2$d %3$s." ), d.beta->disp_name(), buyer_has,
```

File: data/json/npcs/tacoma_ranch/NPC_ranch_bartender.json [hunk 5]
```diff
@@ -48,25 +48,21 @@
       {
         "text": "[$8] I'll take a beer",
```
Comment (Night-Pryanik) [inline data/json/npcs/tacoma_ranch/NPC_ranch_bartender.json]: I guess we need to rename this to mention Merch and not dollars.

**Comment by mlangsdorf** [comment]
jenkins rebuild

**Comment by mlangsdorf** [comment]
jenkins rebuild

File: data/json/npcs/tacoma_ranch/NPC_ranch_bartender.json (hunk 5 above)
Comment (mlangsdorf) [reply data/json/npcs/tacoma_ranch/NPC_ranch_bartender.json]: He'll take $8 worth of barter.  3 Merch will do the job, but you can pay him with other stuff, so I'd prefer it stays $.

**Comment by Fris0uman** [comment]
Are the cash card themselves worth something ?

**Comment by kevingranade** [comment]
They're pretty much plastic scrap, which is nearly worthless.

**Comment by Fris0uman** [comment]
They might also be worth some gas/snack/drinks in vending machines, so not much but something.

**Comment by mlangsdorf** [comment]
#31289 that includes all of this and is merged, so closing this.

**Pull request closed without merging**